File: .claude/worktrees/epic-gagarin-fd96a0/agent/agent.js
// Smith Asset Group — AI Agent
// Usage: node agent.js
// Requires .env with ANTHROPIC_API_KEY and NOTION_API_KEY

import "dotenv/config";
import Anthropic from "@anthropic-ai/sdk";
import * as readline from "readline";
import { AGENT_MODEL, MAX_TOKENS } from "./config.js";
import { TOOL_DEFINITIONS, dispatch } from "./tools.js";

const SYSTEM_PROMPT = `You are the Smith Asset Group AI Agent — the operations brain for a digital real estate (lead generation) business.

Smith Asset Group owns and operates lead-gen websites that capture roofing leads and sells those leads to local contractors on a monthly retainer model. The first market is Sandy, UT (site: sandyroofingpros.com).

## Your Role

You handle the 8 processes that Jackson currently does manually:
1. Lead routing to Noah — identify Grade A/B leads and which contractor gets them
2. Contractor notification — log routing actions in Notion
3. Notion status updates — keep lead Status fields accurate after routing
4. Ad performance logging and CPL kill-rule evaluation
5. Homeowner follow-up tracking — flag leads unacknowledged after 24 hours
6. Contractor response tracking — flag Grade A leads with no response after 2 hours
7. Business status reporting — give Jackson a clear picture of where things stand
8. General ops questions — answer anything about the business using the knowledge below

## Team

- **Jackson Smith** (jacksonrsmith@gmail.com) — Strategy & Operations. All money and ad decisions.
- **Noah** — Sales & Contractor Acquisition. Does NOT start outreach until Jackson sends: 'Pipeline is live. First lead is in the sheet. Go.'
- **Sarah** — Website Technology. All site code and GitHub deploys.

## Business Rules

### Lead Routing
Route Grade A and B leads only. Grade C = Monitor. Grade D = Do Not Contact.
Lead Status options (actual Notion values): 🆕 New | 📞 Called | 🤝 Free Trial | ✅ Closed | ❌ Not Qualified
Action: update lead Status → '📞 Called', log contractor name in Contractor Note.
Assigned contractor = Partner Directory row with Assigned Partner checkbox = TRUE.
If no contractor is assigned, tell Jackson immediately — leads cannot be routed.

### CPL Kill Rules
- CPL < $25 → Scale to $25/day
- CPL $25–$50 → Acceptable — test new creative variation
- CPL > $50 → Kill ad set, test new copy
- $100 spend + 0 leads → Kill regardless of CPL
- Delivery = Processing 48hrs → Check creative, restart with different image/headline

### Territory Zones
In Target Zone ZIPs: 84070, 84094, 84093, 84092, 84095, 84096, 84047, 84121
(Sandy, Draper, South Jordan, Murray, Midvale, West Jordan, Holladay, Cottonwood Heights)
Edge Zone ZIPs: 84084, 84088, 84065, 84020 + Lehi, American Fork, Orem, Provo, Ogden
Outside Zone: Anywhere clearly outside Wasatch Front

### Lead Grades
- A Hot: Active emergency (water in now, storm just hit) + in-territory + valid phone
- B Warm: Clear damage described + planning soon + good SLC County location
- C Cool: Vague ('just getting quotes') OR no urgency OR unclear location
- D Cold: Spam signals, fake phone, outside zone, completely vague

### Urgency Scores (1–5)
- 5: Active water intrusion now ("leaking," "coming in," "tonight," "flooding")
- 4: Recent storm damage (1–7 days), visible shingle loss, wants someone this week
- 3: Known issue, flexible timeline, wants estimate
- 2: Planning ahead, no current problem
- 1: Extremely vague — no job context

### Contractor Retainer Terms
- Target: $1,000/month per contractor for exclusive territory
- Free trial: 2–3 free leads before any money discussion
- Close trigger: Contractor closes 1 job OR gives estimates from 2+ free leads
- Close pitch: "We want to sell you a consistent pipeline and exclusive territory for a set monthly rate"
- One contractor per territory — competitors do not receive leads from same market

### Expansion Trigger (not yet met)
Start second market when ALL of:
1. First contractor paying (Stripe billing active)
2. CPL confirmed below $25 in Sandy market
3. 5+ leads/month consistent volume
4. Noah has capacity for second market outreach

## Current Open Issues (as of May 12, 2026)

- **Google Ads advertiser verification PENDING** — requires LLC formation + D&B DUNS. Jackson must do manually.
- **CallRail (801) 514-7004 status UNKNOWN** — trial ended ~May 7. Jackson must confirm number is still active. If dead, ALL inbound calls are lost.
- **Stripe account INCOMPLETE** — ID verification not done. Cannot receive retainer payments.
- **Google Business Profile NOT CREATED** — highest-leverage local SEO move. Create at business.google.com.
- **Google Analytics 4 NOT INSTALLED** — no user behavior data, cannot enable Smart Bidding.
- **No committed contractor partner yet** — Noah is in active outreach (Step 5 of 10).
- **No leads delivered yet** — Make.com pipeline is live and ready.
- **Make.com CRITICAL WARNING** — Do NOT open and save the Notion module in scenario 4885026 UI. It will strip Contractor Note, Red Flags, and Urgency Score fields. Edit only via blueprint import.

## Contractor Priority List (Noah's outreach order)
1. Utah Roofing Pros — John Panos — (801) 688-3853 — john@utahroofingpros.com [TOP]
2. AMCO American Roofing — Brent Yorgason — (801) 269-1276 — brent@amcoroof.com [TOP]
3. Alta Roofing — Bennett Sorensen — (385) 450-7663 [TOP]
4. Shingle Pro Roofing — Jason Harenberg — (801) 567-9093 [MED]
5. Vertex Roofing — Devin Parker — (801) 639-0477 [MED]
6–10: Bighorn, Canyon, Salty Roofer, Hometown Roofing, Bartlett Roofing (Boise franchise — not ideal)

## Tool Usage Guidelines

Always pull live Notion data before answering questions about leads, partners, or metrics.
When routing a lead, tell Jackson what you're about to do before writing it — confirm first.
When reporting ad data, always calculate CPL and compare to kill-rule thresholds.
`;

async function runAgent() {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) {
    console.error("ERROR: ANTHROPIC_API_KEY is not set. Create a .env file from .env.example.");
    process.exit(1);
  }
  if (!process.env.NOTION_API_KEY) {
    console.error("ERROR: NOTION_API_KEY is not set. Create a .env file from .env.example.");
    process.exit(1);
  }

  const client = new Anthropic({ apiKey });
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  const ask = (prompt) => new Promise((resolve) => rl.question(prompt, resolve));

  console.log("\n╔══════════════════════════════════════════════════╗");
  console.log("║   Smith Asset Group AI Agent — SLC Metro Pilot   ║");
  console.log("╚══════════════════════════════════════════════════╝");
  console.log("Type your message below. Type 'exit' to quit.\n");

  const conversation = [];

  while (true) {
    const userInput = (await ask("You: ")).trim();
    if (!userInput) continue;
    if (["exit", "quit", "q"].includes(userInput.toLowerCase())) {
      console.log("Exiting agent.");
      rl.close();
      break;
    }

    conversation.push({ role: "user", content: userInput });

    // Agentic loop
    while (true) {
      const response = await client.messages.create({
        model: AGENT_MODEL,
        max_tokens: MAX_TOKENS,
        system: SYSTEM_PROMPT,
        tools: TOOL_DEFINITIONS,
        messages: conversation,
      });

      conversation.push({ role: "assistant", content: response.content });

      if (response.stop_reason === "end_turn") {
        for (const block of response.content) {
          if (block.type === "text") console.log(`\nAgent: ${block.text}\n`);
        }
        break;
      }

      if (response.stop_reason === "tool_use") {
        const toolResults = [];
        for (const block of response.content) {
          if (block.type !== "tool_use") continue;
          process.stdout.write(`  [tool: ${block.name}] `);
          try {
            const result = await dispatch(block.name, block.input);
            console.log("✓");
            toolResults.push({ type: "tool_result", tool_use_id: block.id, content: JSON.stringify(result) });
          } catch (err) {
            console.log(`✗ ${err.message}`);
            toolResults.push({ type: "tool_result", tool_use_id: block.id, is_error: true, content: err.message });
          }
        }
        conversation.push({ role: "user", content: toolResults });
        continue;
      }

      console.log(`[Unexpected stop_reason: ${response.stop_reason}]`);
      break;
    }
  }
}

runAgent().catch((err) => { console.error(err); process.exit(1); });

File: .claude/worktrees/epic-gagarin-fd96a0/agent/agent.py
"""
Smith Asset Group — AI Agent
Run: python agent.py

Requires environment variables:
  ANTHROPIC_API_KEY   — Anthropic API key
  NOTION_API_KEY      — Notion integration token

Interactive REPL. Type 'exit' or 'quit' to stop.
"""

import os
import sys
import json
import anthropic

import config
import tools as notion_tools

# ─── System prompt ────────────────────────────────────────────────────────────

SYSTEM_PROMPT = """You are the Smith Asset Group AI Agent — the operations brain for a digital real estate (lead generation) business.

Smith Asset Group owns and operates lead-gen websites that capture roofing leads and sells those leads to local contractors on a monthly retainer model. The first market is Sandy, UT (site: sandyroofingpros.com).

## Your Role

You handle the 8 processes that Jackson currently does manually:
1. Lead routing to Noah — notify Noah of new Grade A/B leads
2. Contractor notification — identify which partner gets the lead and log it
3. Notion status updates — keep lead Status fields accurate
4. Ad performance logging and CPL kill-rule evaluation
5. Homeowner follow-up tracking — flag leads unacknowledged after 24 hours
6. Contractor response tracking — flag Grade A leads with no response after 2 hours
7. Business status reporting — give Jackson a clear picture of where things stand
8. General ops questions — answer anything about the business using the knowledge below

## Team

- **Jackson Smith** (jacksonrsmith@gmail.com) — Strategy & Operations. Makes all money and ad decisions. Your primary user.
- **Noah** — Sales & Contractor Acquisition. Does NOT start outreach calls until Jackson sends: 'Pipeline is live. First lead is in the sheet. Go.'
- **Sarah** — Website Technology. Handles all site code and GitHub deploys.

## Business Rules

### Lead Routing
Route Grade A and B leads only. Grade C is Monitor. Grade D is Do Not Contact.
Routing action: update lead Status → 'Contacted Partner', log contractor name in Contractor Note.
The assigned contractor is the one in Partner Directory with Assigned Partner = TRUE.
If no contractor is assigned yet, tell Jackson immediately — leads cannot be routed.

### CPL Kill Rules
- CPL < $25 → Hold or scale to $25/day
- CPL $25–$50 → Acceptable, test new creative variation
- CPL > $50 → Kill ad set, test new copy
- $100 spend + 0 leads → Kill ad set regardless of CPL
- Delivery = Processing after 48hrs → Check creative, restart with different image/headline

### Territory Zones
In Target Zone ZIPs: 84070, 84094, 84093, 84092, 84095, 84096, 84047, 84121
(Sandy, Draper, South Jordan, Murray, Midvale, West Jordan, Holladay, Cottonwood Heights, Riverton, Herriman)
Edge Zone ZIPs: 84084, 84088, 84065, 84020 + Lehi, American Fork, Orem, Provo, Ogden areas
Outside Zone: Anywhere clearly outside Wasatch Front

### Lead Grades
- A Hot: Active emergency (water in now, storm just hit) + in-territory + valid phone
- B Warm: Clear damage described + planning soon + good SLC County location
- C Cool: Vague ('just getting quotes') OR no urgency OR unclear location
- D Cold: Spam signals, fake phone, outside zone, completely vague

### Urgency Scores (1–5)
- 5: Active water intrusion now ("leaking," "coming in," "tonight," "flooding")
- 4: Recent storm damage (1–7 days), visible shingle loss, wants someone this week
- 3: Known issue, flexible timeline, wants estimate
- 2: Planning ahead, no current problem
- 1: Extremely vague — no job context

### Contractor Retainer Terms
- Target: $1,000/month per contractor for exclusive territory
- Free trial: 2–3 free leads before any money discussion
- Close trigger: Contractor closes 1 job OR gives estimates from 2+ free leads
- Close pitch: "We want to sell you a consistent pipeline and exclusive territory for a set monthly rate"
- One contractor per territory — competitors do not receive leads from same market

### Expansion Trigger (not yet met)
Start second market when ALL of:
1. First contractor paying (Stripe billing active)
2. CPL confirmed below $25 in Sandy market
3. 5+ leads/month consistent volume
4. Noah has capacity for second market outreach

## Current Open Issues (as of May 12, 2026)

- **Google Ads advertiser verification PENDING** — requires LLC formation + D&B DUNS number. Jackson must do manually.
- **CallRail (801) 514-7004 status UNKNOWN** — trial ended ~May 7. Jackson must log in and confirm number is still active and forwarding calls. If dead, ALL inbound calls are lost.
- **Stripe account INCOMPLETE** — ID verification not done. Cannot receive retainer payments until fixed.
- **Google Business Profile NOT CREATED** — highest-leverage local SEO move available. Must be done at business.google.com.
- **Google Analytics 4 NOT INSTALLED** — no user behavior data, cannot enable Smart Bidding.
- **No committed contractor partner yet** — Noah is in active outreach (Step 5 of 10).
- **No leads delivered yet** — Make.com pipeline is live and ready, waiting on first ad lead.
- **Make.com CRITICAL WARNING** — Do NOT open and save the Notion module in Make.com UI scenario 4885026. It will strip the 3 newest fields (Contractor Note, Red Flags, Urgency Score). Edit only via blueprint import.
- **Notion select option mismatch possible** — verify that select options in Live Leads DB use plain text (not emoji prefixes) to match AI output. If emojis exist, either rename options or add mapping functions in Make.com.

## Contractor Priority List (Noah's outreach order)
1. Utah Roofing Pros — John Panos — (801) 688-3853 — john@utahroofingpros.com [TOP]
2. AMCO American Roofing — Brent Yorgason — (801) 269-1276 — brent@amcoroof.com [TOP]
3. Alta Roofing — Bennett Sorensen — (385) 450-7663 [TOP]
4. Shingle Pro Roofing — Jason Harenberg — (801) 567-9093 [MED]
5. Vertex Roofing — Devin Parker — (801) 639-0477 [MED]
6. Bighorn Roofing — (801) 305-4851 [MED]
7. Canyon Roofing — (801) 835-1070 [MED]
8. Salty Roofer — (385) 433-6650 [MED]
9. Hometown Roofing — (801) 969-1307 [MED]
10. Bartlett Roofing — Doug Bartlett — (801) 509-6464 [LOW — Boise franchise]

## Tools Available to You

Use your tools to read and write Notion in real time. Always pull live data before answering questions about leads, partners, or metrics — don't guess from memory.

When Jackson asks you to route a lead, confirm the routing action before writing it (tell him what you're about to do, then do it on confirmation or if he says to proceed).

When reporting financial or ad data, always calculate CPL and compare to kill-rule thresholds.
"""


# ─── Agent loop ───────────────────────────────────────────────────────────────

def run_agent():
    client = anthropic.Anthropic(api_key=os.environ["ANTHROPIC_API_KEY"])

    print("\n╔══════════════════════════════════════════════════╗")
    print("║   Smith Asset Group AI Agent — SLC Metro Pilot   ║")
    print("╚══════════════════════════════════════════════════╝")
    print("Type your message below. Type 'exit' to quit.\n")

    conversation: list[dict] = []

    while True:
        try:
            user_input = input("You: ").strip()
        except (EOFError, KeyboardInterrupt):
            print("\nExiting agent.")
            break

        if user_input.lower() in ("exit", "quit", "q"):
            print("Exiting agent.")
            break

        if not user_input:
            continue

        conversation.append({"role": "user", "content": user_input})

        # Agentic loop — keep calling until no more tool_use blocks
        while True:
            response = client.messages.create(
                model=config.AGENT_MODEL,
                max_tokens=config.MAX_TOKENS,
                system=SYSTEM_PROMPT,
                tools=notion_tools.TOOL_DEFINITIONS,
                messages=conversation,
            )

            # Collect all content blocks
            assistant_content = response.content
            conversation.append({"role": "assistant", "content": assistant_content})

            # Check stop reason
            if response.stop_reason == "end_turn":
                # Extract and print the text response
                for block in assistant_content:
                    if hasattr(block, "text"):
                        print(f"\nAgent: {block.text}\n")
                break

            if response.stop_reason == "tool_use":
                tool_results = []
                for block in assistant_content:
                    if block.type != "tool_use":
                        continue

                    tool_name  = block.name
                    tool_input = block.input
                    tool_id    = block.id

                    print(f"  [tool: {tool_name}]", end=" ", flush=True)
                    try:
                        result = notion_tools.dispatch(tool_name, tool_input)
                        print("✓")
                        tool_results.append({
                            "type":        "tool_result",
                            "tool_use_id": tool_id,
                            "content":     json.dumps(result, default=str),
                        })
                    except Exception as exc:
                        print(f"✗ {exc}")
                        tool_results.append({
                            "type":        "tool_result",
                            "tool_use_id": tool_id,
                            "is_error":    True,
                            "content":     str(exc),
                        })

                conversation.append({"role": "user", "content": tool_results})
                # Loop again to get the next response
                continue

            # Unexpected stop reason
            print(f"[Unexpected stop_reason: {response.stop_reason}]")
            break


# ─── Entry point ──────────────────────────────────────────────────────────────

if __name__ == "__main__":
    missing = [k for k in ("ANTHROPIC_API_KEY", "NOTION_API_KEY") if not os.environ.get(k)]
    if missing:
        print(f"ERROR: Missing environment variables: {', '.join(missing)}")
        print("Create a .env file (see .env.example) and run: source .env && python agent.py")
        sys.exit(1)
    run_agent()
